x-pack/plugins/security_solution/public/management/pages/trusted_apps/store/middleware.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ServerApiError } from '../../../../common/types';
+import { Immutable } from '../../../../../common/endpoint/types';
+import { ImmutableMiddlewareAPI, ImmutableMiddlewareFactory } from '../../../../common/store';
+import { AppAction } from '../../../../common/store/actions';
+import {
+  AsyncDataBinding,
+  StaleAsyncBinding,
+  getLastPresentDataBinding,
+} from '../state/async_data_binding';
+import {
+  TrustedAppsListData,
+  TrustedAppsListPageState,
+} from '../state/trusted_apps_list_page_state';
+import { TrustedAppsHttpService, TrustedAppsService } from '../service';
+import { needsRefreshOfListData } from './selectors';
+import { TrustedAppsListDataBindingChanged } from './action';
+
+const createListDataBindingChangedAction = (
+  newBinding: Immutable<AsyncDataBinding<TrustedAppsListData, ServerApiError>>
+): Immutable<TrustedAppsListDataBindingChanged> => ({
+  type: 'trustedAppsListDataBindingChanged',
+  payload: { newBinding },
+});
+
+const refreshList = async (
+  store: ImmutableMiddlewareAPI<TrustedAppsListPageState, AppAction>,
+  trustedAppsService: TrustedAppsService
+) => {
+  const list = store.getState().listView;
+
+  store.dispatch(
+    createListDataBindingChangedAction({
+      type: 'InProgressAsyncBinding',
+      // need to think on how to avoid the casting
+      previousBinding: list.currentListData as Immutable<
+        StaleAsyncBinding<TrustedAppsListData, ServerApiError>
+      >,
+    })
+  );
+
+  try {
+    const response = await trustedAppsService.getTrustedAppsList({
+      page: list.currentPaginationInfo.index + 1,
+      per_page: list.currentPaginationInfo.size,
+    });
+
+    store.dispatch(
+      createListDataBindingChangedAction({
+        type: 'PresentAsyncBinding',
+        data: {
+          items: response.data,
+          paginationInfo: list.currentPaginationInfo,
+          totalItemsCount: response.total,
+        },
+      })
+    );
+  } catch (error) {
+    store.dispatch(
+      createListDataBindingChangedAction({
+        type: 'FailedAsyncBinding',
+        error,
+        lastPresentBinding: getLastPresentDataBinding(list.currentListData),
+      })
+    );
+  }
+};
+
+export const trustedAppsPageMiddlewareFactory: ImmutableMiddlewareFactory<TrustedAppsListPageState> = (
```

Comment:
Question (could be addressed in another PR), what happens when we leave the Trusted Apps view?  Is state cleared?  When we enter back in, which state will we be in, ( i.e.  `uninitialized`, `loading`, `success` ?).  Endpoint list currently does a bad job at this, so I wouldn't block the PR, just something to continue thinking about.  cc @paul-tavares @parkiino

Reply:
It depends :P

First of all - no, state is not cleared, last state is kept, but...

The state of the page when you come back is depending on location that you come back to. So when you just use history to come back everything is already there and in the right state. When you come back to a different location than the one you left the state is updated, for example:
   * you opened the trusted apps tab
   * you went on the page 5
   * you opened the endpoints tab
   * you opened the trusted apps tab again - your location from URI is now page 1 so you will have data reloaded

But now you made me think of that case and yes there will be some weirdness - when you do the step 4 from the above steps you'll see data for page 5 for a short amount of time cause that's a previous state. As a quick fix I see two options:
   * When location moves outside trusted apps reset the binding/resource with uninitialised state. This has a side effect that history navigation will trigger loading when it could skip it (probably minor), but it's very simple to implement.
   * Not use the table loading indicator and not show the previous items but every time location change empty table and load items while showing some spinner. Pretty ugly for pagination.

For now the first option is very simple and has more or less nice interaction and nicely fit with empty page state requirements and lets discuss this more in detail with empty state story cause there are quite many approaches to it each with it's own drawbacks. I think main thing to do before choosing implementation strategy is to define UX requirements that we want to achieve and clearly define test cases.

Reply:
I've implemented option 1 (resetting state on navigating away) and let's revisit this when implementing empty state story.